File: src/main/java/com/leijendary/spring/microservicetemplate/error/DataIntegrityViolationExceptionHandler.java
package com.leijendary.spring.microservicetemplate.error;

import com.leijendary.spring.microservicetemplate.data.response.ErrorResponse;
import lombok.RequiredArgsConstructor;
import org.hibernate.exception.ConstraintViolationException;
import org.springframework.context.MessageSource;
import org.springframework.core.annotation.Order;
import org.springframework.dao.DataIntegrityViolationException;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;

import static com.leijendary.spring.microservicetemplate.util.RequestContext.getLocale;
import static com.leijendary.spring.microservicetemplate.util.StringUtil.snakeCaseToCamelCase;
import static org.apache.commons.lang3.StringUtils.substringBetween;
import static org.springframework.http.HttpStatus.CONFLICT;
import static org.springframework.http.HttpStatus.INTERNAL_SERVER_ERROR;

@RestControllerAdvice
@Order(3)
@RequiredArgsConstructor
public class DataIntegrityViolationExceptionHandler {

    private final MessageSource messageSource;

    @ExceptionHandler(DataIntegrityViolationException.class)
    public ResponseEntity<ErrorResponse> catchDataIntegrityViolation(final DataIntegrityViolationException exception) {
        final var errorResponse = getResponse(exception);
        final var status = errorResponse.getMeta().get("status");

        return ResponseEntity
                .status((int) status)
                .body(errorResponse);
    }

    private ErrorResponse getResponse(final DataIntegrityViolationException exception) {
        final var cause = exception.getCause();

        if (cause instanceof ConstraintViolationException) {
            return constraintViolationException((ConstraintViolationException) cause);
        }

        final var code = "error.dataIntegrity";
        final var message = messageSource.getMessage(code, new Object[] { exception.getMessage() }, getLocale());

        return ErrorResponse.builder()
                .addError("Entity", code, message)
                .status(INTERNAL_SERVER_ERROR)
                .build();
    }

    private ErrorResponse constraintViolationException(final ConstraintViolationException exception) {
        final var errorMessage = exception.getSQLException().getNextException().getMessage();
        final var field = snakeCaseToCamelCase(substringBetween(errorMessage, "Key (", ")="));
        final var value = substringBetween(errorMessage, "=(", ") ");
        final var code = "validation.alreadyExists";
        final var message = messageSource.getMessage(code, new Object[] { field, value }, getLocale());

        return ErrorResponse.builder()
                .addError(field, code, message)
                .status(CONFLICT)
                .build();
    }
}

File: src/main/kotlin/com/leijendary/spring/template/core/error/DataIntegrityViolationExceptionHandler.kt
package com.leijendary.spring.template.core.error

import com.leijendary.spring.template.core.extension.snakeCaseToCamelCase
import com.leijendary.spring.template.core.model.ErrorModel
import com.leijendary.spring.template.core.util.RequestContext.locale
import org.hibernate.exception.ConstraintViolationException
import org.postgresql.util.PSQLException
import org.springframework.context.MessageSource
import org.springframework.core.annotation.Order
import org.springframework.dao.DataIntegrityViolationException
import org.springframework.http.HttpStatus
import org.springframework.http.HttpStatus.CONFLICT
import org.springframework.http.HttpStatus.INTERNAL_SERVER_ERROR
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice
import java.sql.BatchUpdateException

@RestControllerAdvice
@Order(3)
class DataIntegrityViolationExceptionHandler(private val messageSource: MessageSource) {
    @ExceptionHandler(DataIntegrityViolationException::class)
    fun catchDataIntegrityViolation(exception: DataIntegrityViolationException): ResponseEntity<List<ErrorModel>> {
        val (errors, status) = errorPair(exception)

        return ResponseEntity
            .status(status)
            .body(errors)
    }

    private fun errorPair(exception: DataIntegrityViolationException): Pair<List<ErrorModel>, HttpStatus> {
        val cause = exception.cause

        if (cause is ConstraintViolationException) {
            return constraint(cause)
        }

        val code = "error.data.integrity"
        val arguments = arrayOf(exception.message ?: "")
        val message = messageSource.getMessage(code, arguments, locale)
        val error = ErrorModel(mutableListOf("data", "entity"), code, message)

        return listOf(error) to INTERNAL_SERVER_ERROR
    }

    private fun constraint(exception: ConstraintViolationException): Pair<List<ErrorModel>, HttpStatus> {
        val sqlException = when (exception.sqlException) {
            is BatchUpdateException -> exception.sqlException.nextException
            else -> exception.sqlException
        } as PSQLException
        val errorMessage = sqlException.serverErrorMessage!!
        val table = errorMessage.table!!.snakeCaseToCamelCase(true)
        val detail = errorMessage.detail!!
        val column = errorMessage.column ?: detail
            .substringAfter("Key (")
            .substringBefore("))=")
            .substringBefore(")=")
            .substringAfter("(")
            .substringBefore("::")
        val field = column.snakeCaseToCamelCase()
        val value = detail.substringAfter("=(").substringBefore(") ")
        val code = "validation.alreadyExists"
        val arguments = arrayOf(field, value)
        val message = messageSource.getMessage(code, arguments, locale)
        val error = ErrorModel(mutableListOf("data", table, field), code, message)

        return listOf(error) to CONFLICT
    }
}
